### DISCUSSION TITLE
How to import TypeScript types such as Scale?

### DISCUSSION THREAD

**Discussion by stevage** [tonaljs/tonal#401] [Q&A]
Hi, I'm not super familiar with TypeScript but using it with Tonal in a Vue project.

I'm a bit confused how to access types such as Scale (ie, the return value of `Scale.get()`), because of the name collision with the module name.

`import { Scale } from 'tonal'`

This gives us `Scale`, the whole module.

There's also the type, `Scale` declared like this in `packages/scale.ts`:

```
export interface Scale extends ScaleType {
  tonic: string | null;
  type: string;
  notes: NoteName[];
}
```

But I can't figure out how to import that `Scale` type in order to declare a variable of type `Scale`.

These don't seem to work:

```
import type { Scale } from 'tonal'
import type { Scale.Scale } from 'tonal'
import type { Scale } from 'tonal/packages/scale'
```

**Comment by danigb** [comment]
The solution is to install `@tonaljs/scale` package (`npm i @tonaljs/scale`) and import from there:

```ts
import { Scale, get } from "@tonaljs/scale";
const myScale: Scale = { ... }
console.log(myScale);
console.log(get("C major")); // <= just a reminder the function from packages can be imported and used without namespace
```

I never realized those types were not exported from main `tonal` package... so I'm curious: why do you need it?

**Reply by bali182** [reply]
Just came across this myself, in case you don't just grab something quick from the return value of a tonal-call, it would be beneficial to be able to refer to the return types, eg.:

```ts
const key = Key.majorKey('C')

const tonics = getTriadsByFunction(key, 'T')
const subdominants= getTriadsByFunction(key, 'SD')
const dominants = getTriadsByFunction(key, 'D')

function getTriadsByFunction(key: Key, chordFunction: ChordFunction): string[] {
  return key.chordsHarmonicFunction
    .map((fn, i) => (fn === 'T' ? key.triads[i] : undefined))
    .filter(isNotNil)
}
```

This is just an example, but IMO it's very useful for any library that has typescript bindings to expose the parameter and return types of their interface.

**Comment by danigb** [comment]
In the new version you can import types directly (`NoteType`, `ScaleType`, ...):
```ts
import { Interval, IntervalType, Note, NoteType } from "tonal";

const note: NoteType = Note.get("C4");
const interval: IntervalType = Interval.get("4P");
```
